File: src/main/java/it/alian/gun/mesmerize/listeners/ItemListener.java
package it.alian.gun.mesmerize.listeners;

import it.alian.gun.mesmerize.MConfig;
import it.alian.gun.mesmerize.Mesmerize;
import it.alian.gun.mesmerize.compat.AttackSpeed;
import it.alian.gun.mesmerize.compat.Compat;
import it.alian.gun.mesmerize.lore.ItemInfo;
import it.alian.gun.mesmerize.lore.LoreInfo;
import it.alian.gun.mesmerize.lore.LoreParser;
import it.alian.gun.mesmerize.util.Math;
import org.bukkit.Bukkit;
import org.bukkit.entity.Player;
import org.bukkit.event.Event;
import org.bukkit.event.EventHandler;
import org.bukkit.event.HandlerList;
import org.bukkit.event.Listener;
import org.bukkit.event.block.BlockBreakEvent;
import org.bukkit.event.entity.EntityDeathEvent;
import org.bukkit.event.entity.PlayerDeathEvent;
import org.bukkit.event.inventory.InventoryClickEvent;
import org.bukkit.event.player.PlayerInteractEvent;
import org.bukkit.event.player.PlayerItemDamageEvent;
import org.bukkit.event.player.PlayerItemHeldEvent;
import org.bukkit.event.player.PlayerPickupItemEvent;
import org.bukkit.inventory.ItemStack;

public class ItemListener implements Listener {

    @EventHandler
    public void onOpenInv(PlayerInteractEvent event) {

    }

    @EventHandler
    public void onDeath(PlayerDeathEvent event) {

    }

    @EventHandler
    public void onBlockBreak(BlockBreakEvent event) {
        long nano = System.nanoTime();
        if (event.getPlayer() != null && event.getExpToDrop() != 0) {
            event.setExpToDrop((int) (event.getExpToDrop() *
                    (LoreParser.getByEntityId(event.getPlayer().getEntityId()).getOtherExpModifier() + 1)));
        }
        if (MConfig.debug)
            System.out.println(event.getEventName() + " processed in " + (System.nanoTime() - nano) * 1E-6 + " ms.");
    }

    @EventHandler
    public void onEntityDeath(EntityDeathEvent event) {
        long nano = System.nanoTime();
        if (event.getEntity().getKiller() != null && event.getDroppedExp() != 0) {
            LoreInfo info = LoreParser.getByEntityId(event.getEntity().getKiller().getEntityId());
            event.setDroppedExp((int) (event.getDroppedExp() * (info.getAttackExpModifier() + 1)));
        }
        if (MConfig.debug)
            System.out.println(event.getEventName() + " processed in " + (System.nanoTime() - nano) * 1E-6 + " ms.");
    }

    @EventHandler
    public void onItemChange(PlayerItemHeldEvent event) {
        long nano = System.nanoTime();
        if (LoreParser.check(event.getPlayer().getInventory().getItem(event.getPreviousSlot()), event.getPlayer())) {
            event.setCancelled(true);
        }
        if (MConfig.debug)
            System.out.println(event.getEventName() + " processed in " + (System.nanoTime() - nano) * 1E-6 + " ms.");
    }

    @EventHandler
    public void onItemUse(PlayerItemDamageEvent event) {
        long nano = System.nanoTime();
        ItemInfo info = LoreParser.parseItem(event.getItem());
        if (Math.random() < info.getUnbreakable()) event.setDamage(0);
        if (event.getItem().getDurability() == (event.getItem().getType().getMaxDurability() - 1) && !MConfig.General.breakOnDurabilityOff) {
            ItemStack itemStack = event.getItem().clone();
            event.getPlayer().getWorld().dropItemNaturally(event.getPlayer().getLocation(), itemStack);
            event.setCancelled(true);
            event.getPlayer().getInventory().remove(itemStack);
            event.getPlayer().sendMessage(String.format(MConfig.Message.onDurabilityItemDrop, itemStack.getItemMeta().getDisplayName()));
        } else if (event.getItem().getType().getMaxDurability() != 0) {
            double prev = ((double) event.getItem().getDurability()) / ((double) event.getItem().getType().getMaxDurability());
            double now = ((double) (event.getItem().getDurability() + event.getDamage())) / ((double) event.getItem().getType().getMaxDurability());
            for (double v : MConfig.General.durabilityWarnThreshold) {
                if (v > prev && v <= now) {
                    event.getPlayer().sendMessage(String.format(MConfig.Message.onDurabilityWarn,
                            event.getItem().hasItemMeta() ? event.getItem().getItemMeta().hasDisplayName() ?
                                    event.getItem().getItemMeta().getDisplayName() : event.getItem().getType().name() :
                                    event.getItem().getType().name(), ((1D - prev) * 100D)));
                }
            }
        }
        if (MConfig.debug)
            System.out.println(event.getEventName() + " processed in " + (System.nanoTime() - nano) * 1E-6 + " ms.");
    }

    @EventHandler
    public void onPickup(PlayerPickupItemEvent event) {
        if (LoreParser.check(event.getItem().getItemStack(), event.getPlayer())) {
            event.setCancelled(true);
        }
    }

    @EventHandler
    public void onClick(InventoryClickEvent event) {
        /*
        if (LoreParser.check(event.getCurrentItem(), event.getWhoClicked())) {
            event.setCancelled(true);
            event.getWhoClicked().sendMessage(MConfig.Message.omSoulboundCheck);
        }
        */
    }

    @EventHandler
    public void onTick(TickEvent event) {
        if (MConfig.General.enableHealthControl)
            for (Player player : Compat.getOnlinePlayers()) {
                if (!player.isDead() && player.isValid()) {
                    double health = Math.min(LoreParser.getByEntityId(player.getEntityId()).getRegeneration() + player.getHealth(),
                            player.getMaxHealth());
                    player.setHealth(health);
                }
            }
    }

    public static void init() {
        Bukkit.getPluginManager().registerEvents(new ItemListener(), Mesmerize.instance);
        Bukkit.getScheduler().runTaskTimer(Mesmerize.instance, () ->
                Bukkit.getPluginManager().callEvent(new TickEvent()), 10, MConfig.General.regenInterval);
        Bukkit.getScheduler().runTaskTimer(Mesmerize.instance, () -> {
            for (Player livingEntity : Compat.getOnlinePlayers()) {
                LoreInfo info = LoreParser.getByEntityId(livingEntity.getEntityId());
                if (MConfig.General.enableHealthControl) {
                    if (MConfig.General.healthScaled) {
                        livingEntity.setHealthScaled(true);
                        livingEntity.setHealthScale(MConfig.General.healthScale);
                    } else {
                        livingEntity.setHealthScaled(false);
                    }
                    livingEntity.setMaxHealth(Math.constraint(MConfig.General.maximumHealth,
                            MConfig.General.minimalHealth, MConfig.General.baseHealth + info.getHealth()));
                }
                livingEntity.setWalkSpeed((float) Math.constraint(0.9999,
                        MConfig.General.minimalMovespeed, MConfig.General.baseMovespeed + info.getMoveSpeed()));
                livingEntity.setFlySpeed((float) Math.constraint(0.999,
                        MConfig.General.minimalFlyspeed, MConfig.General.baseFlyspeed + info.getFlySpeed()));
                AttackSpeed.set(livingEntity, MConfig.General.baseAttackSpeed + info.getAttackSpeed());
            }
        }, 5, MConfig.Performance.loreUpdateInterval);
    }

    public static class TickEvent extends Event {

        private static final HandlerList HANDLER_LIST = new HandlerList();

        @Override
        public HandlerList getHandlers() {
            return HANDLER_LIST;
        }

        public static HandlerList getHandlerList() {
            return HANDLER_LIST;
        }

    }

}

File: src/main/scala/it/alian/gun/mesmerize/listener/ItemListener.scala
package it.alian.gun.mesmerize.listener

import it.alian.gun.mesmerize.MesmerizeDelegate._
import it.alian.gun.mesmerize.compat.Compat
import it.alian.gun.mesmerize.lore.LoreParser
import it.alian.gun.mesmerize.scalaapi.Prelude._
import org.bukkit.entity.Player
import org.bukkit.event.block.BlockBreakEvent
import org.bukkit.event.entity.{EntityDeathEvent, EntityPickupItemEvent, PlayerDeathEvent}
import org.bukkit.event.inventory.InventoryClickEvent
import org.bukkit.event.inventory.InventoryType.SlotType
import org.bukkit.event.player.{PlayerInteractEvent, PlayerItemDamageEvent, PlayerItemHeldEvent}
import org.bukkit.event.{EventHandler, Listener}
import org.bukkit.inventory.ItemStack

import scala.collection.JavaConverters._

object ItemListener extends Listener {

  private def check(stack: ItemStack, player: Player): Boolean = {
    val info = LoreParser.parse(stack)
    if (info.str("soulbound").nonEmpty && info.str("soulbound") != player.getName) {
      runTaskAsync(player.sendMessage(config("message.omSoulboundCheck", "")))
      return false
    }
    if (info.num("levelCap") > player.getLevel) {
      runTaskAsync(player.sendMessage(config("message.onLevelCheck", "").format(info.num("levelCap"))))
      return false
    }
    val cap = info.str("permissionCap")
    if (cap.nonEmpty && !player.hasPermission(config("general.permissionAlias." + cap, cap))) {
      runTaskAsync(player.sendMessage(config("message.onPermissionCheck", "").format(cap)))
      return false
    }
    true
  }

  @EventHandler
  def onOpenInv(event: PlayerInteractEvent): Unit = {
  }

  @EventHandler
  def onDeath(event: PlayerDeathEvent): Unit = {
  }

  @EventHandler
  def onBlockBreak(event: BlockBreakEvent): Unit = {
    if (event.getPlayer != null && event.getExpToDrop != 0)
      event.setExpToDrop((event.getExpToDrop * (LoreParser.parse(event.getPlayer).num("otherExpModifier") + 1)).toInt)
  }

  @EventHandler
  def onEntityDeath(event: EntityDeathEvent): Unit = {
    if (event.getEntity.getKiller != null && event.getDroppedExp != 0)
      event.setDroppedExp((event.getDroppedExp * (LoreParser.parse(event.getEntity.getKiller).num("attackExpModifier") + 1)).toInt)
  }

  @EventHandler
  def onItemChange(event: PlayerItemHeldEvent): Unit = {
    runTaskAsync(LoreParser.update(event.getPlayer): Unit)
    if (!event.getPlayer.getInventory.getItem(event.getNewSlot).empty && !check(event.getPlayer.getInventory.getItem(event.getNewSlot), event.getPlayer))
      event.setCancelled(true)
  }

  @EventHandler
  def onItemUse(event: PlayerItemDamageEvent): Unit = {
    val info = LoreParser.parse(event.getItem)
    if (math.random < info.num("unbreakable")) event.setDamage(0)
    if (event.getItem.getDurability == (event.getItem.getType.getMaxDurability - 1) && !config("general.breakOnDurabilityOff", true)) {
      val itemStack = event.getItem.clone
      event.getPlayer.getWorld.dropItemNaturally(event.getPlayer.getLocation, itemStack)
      event.setCancelled(true)
      event.getPlayer.getInventory.remove(itemStack)
      runTaskAsync(event.getPlayer.sendMessage(config("message.onDurabilityItemDrop", "").format(itemStack.getItemMeta.getDisplayName)))
    } else if (event.getItem.getType.getMaxDurability != 0) {
      val prev = event.getItem.getDurability.toDouble / event.getItem.getType.getMaxDurability.toDouble
      val now = (event.getItem.getDurability + event.getDamage).toDouble / event.getItem.getType.getMaxDurability.toDouble
      for (v <- config.getDoubleList("general.durabilityWarnThreshold").asScala if v > prev && v <= now) {
        event.getPlayer.sendMessage(config("message.onDurabilityWarn", "").format(
          if (event.getItem.hasItemMeta)
            if (event.getItem.getItemMeta.hasDisplayName) event.getItem.getItemMeta.getDisplayName
            else event.getItem.getType.name
          else event.getItem.getType.name, (1D - prev) * 100D))
      }
    }
  }

  @EventHandler
  def onPickup(event: EntityPickupItemEvent): Unit = event.getEntity match {
    case player: Player => if (!check(event.getItem.getItemStack, player)) event.setCancelled(true)
    case _ =>
  }

  @EventHandler
  def onClick(event: InventoryClickEvent): Unit = {
    if (event.getRawSlot == event.getWhoClicked.getInventory.getHeldItemSlot || event.getSlotType == SlotType.ARMOR)
      runTaskAsync(LoreParser.update(event.getWhoClicked): Unit)
  }

  runTask(10, config.getLong("general.regenInterval", 10)) {
    for (player <- Compat.getOnlinePlayers if !player.isDead && player.isValid) {
      val health = (LoreParser.parse(player).num("regeneration") + player.getHealth) min player.getMaxHealth
      player.setHealth(health)
    }
  }

  runTask(5, config.getLong("performance.loreUpdateInterval", 5)) {
    for (livingEntity <- Compat.getOnlinePlayers) {
      val info = LoreParser.parse(livingEntity)
      if (config("general.enableHealthControl", true)) {
        if (config("general.healthScaled", true)) {
          livingEntity.setHealthScaled(true)
          livingEntity.setHealthScale(config("general.healthScale", 20D))
        } else livingEntity.setHealthScaled(false)
        livingEntity.setMaxHealth(config("general.maximumHealth", 20D) min (config("prefix.health.base", 20D) + info.num("health")) max config("general.minimalHealth", 1D))
      }
      livingEntity.setWalkSpeed((0.9999 min (config("prefix.moveSpeed.base", 0.2) + info.num("moveSpeed")) max config("general.minimalMovespeed", 0.0001)).toFloat)
      livingEntity.setFlySpeed((0.9999 min (config("prefix.flySpeed.base", 0.2) + info.num("flySpeed")) max config("general.minimalFlyspeed", 0.0001)).toFloat)
      livingEntity.setAttackSpeed(config("prefix.attackSpeed.base", 18D) + info.num("attackSpeed"))
    }
  }

}